src/main/scala/gtirb/MapDecoder.scala

```diff
@@ -0,0 +1,112 @@
+
+
+
+/* 
+* Provides some useful decoders for certian AuxData sections in gtirb.
+*
+* See https://grammatech.github.io/gtirb/python/_modules/gtirb/serialization.html#MappingCodec.decode, this was pulled
+* from their python API, and converted into scala 
+*/
+object MapDecoder {
+
+  def decode_set(totalBytes: Seq[ByteString]): mutable.Map[ByteString, mutable.Set[ByteString]]  = {
+
+    val totalMap: mutable.Map[ByteString, mutable.Set[ByteString]]
+    = mutable.Map.empty[ByteString, mutable.Set[ByteString]]
+
+    for (bytes <- totalBytes) {
+      val bytesArr: Array[Byte] = bytes.toByteArray
+      val byteStream = ByteArrayInputStream(bytesArr)
+
+      val map = mutable.Map.empty[ByteString, mutable.Set[ByteString]]
+
+      val len = bytesToInt(read_bytes(8, byteStream), true)
+      val num = len.toInt
+      for (s <- 0 until num) {
+        val key = ByteString.copyFrom(read_bytes(16, byteStream))
+        val uuids = mutable.Set[ByteString]();
+        val len = bytesToInt(read_bytes(8, byteStream), true);
+        for (k <- 0 until len.toInt) {
+          val byte = ByteString.copyFrom(read_bytes(16, byteStream))
+          uuids += byte
+        }
+
+        map += (key -> uuids)
+      }
+      totalMap ++= map
+    }
+    totalMap
+
+  }
+
+  def decode_uuid(totalBytes: Seq[ByteString]): mutable.Map[ByteString, ByteString]  = {
+
+    val totalMap : mutable.Map[ByteString, ByteString]
+    = mutable.Map.empty[ByteString, ByteString]
+
+    for (bytes <- totalBytes) {
+      val bytesArr: Array[Byte] = bytes.toByteArray
+      val byteStream = ByteArrayInputStream(bytesArr)
+
+      val map = mutable.Map.empty[ByteString, ByteString]
+
+      val len = bytesToInt(read_bytes(8, byteStream), true)
+      val num = len.toInt
+      for (s <- 0 until num) {
+        val key = ByteString.copyFrom(read_bytes(16, byteStream))
+        val uuid = ByteString.copyFrom(read_bytes(16, byteStream))
+        map += (key -> uuid);
+      }
+      totalMap ++= map
+    }
+    totalMap
+  }
+
+  def decode_string(totalBytes: Seq[ByteString]): mutable.Map[ByteString, String] = {
+    // THIS DOESNT WORK YET
+    // literally can't figure out what's wrong, might be something to do with java/scala treating all bits as signed,
+    // when api is unsigned
+    val totalMap = mutable.Map.empty[ByteString, String]
+
+    for (bytes <- totalBytes) {
+
+      val map = mutable.Map.empty[ByteString, String]
+      val bytesArr: Array[Byte] = bytes.toByteArray
+      val byteStream = ByteArrayInputStream(bytesArr)
+
+      val len = bytesToInt(read_bytes(8, byteStream), true)
+      val num = len.toInt
+      for (s <- 0 until num) {
+        val key = ByteString.copyFrom(read_bytes(16, byteStream))
+        val len = bytesToInt(read_bytes(8, byteStream), true)
+        val str = String(read_bytes(len.toInt, byteStream), StandardCharsets.UTF_8)
+        map += (key -> str)
+      }
+      totalMap ++= map
+    }
+    totalMap
+  }
```

Comment:
This isn't used yet and is listed as not working, what is it supposed to do?

Reply:
This connects to the function argument/ typing problem you mention down below. Gtirb has an algorthim for inferring types known as retypd. The problem is that their plugin to use this algorithm in ddisasm is only for x86, and also isn't maintained and doesn't work with the current version of ddiasm. We can implement the plugin ourselves, but that requires some data from that static analysis later in basil that isn't quite done yet. 

Thas was created before I knew the above problem. The "types" auxdata section has mapping uuid -> string, so I figured I'd create a decoder and test it on some other audata section, but that fell through. So i think it's largely useless now.